src/http_server.c

```diff
@@ -0,0 +1,150 @@
+#include <netinet/in.h>
+#include <string.h>
+#include <unistd.h>
+#include <ctype.h>
+#include <netinet/tcp.h>
+
+#include "config.h"
+#include "log.h"
+#include "tunnel.h"
+
+// Convenience function to send a response with a user readable status message and the
+// request URL shown for debug purposes.
+// The response is shown in the user's browser after being redirected from the Fortinet Server.
+static void send_status_response(int socket, const char *userMessage) {
+    const char *reply = "HTTP/1.1 200 OK\r\n"
+            "Content-Type: text/plain\r\n"
+            "Content-Length: %lu\r\n"
+            "Connection: close\r\n\r\n"
+            "%s"; // User readable message
+
+    int requiredSize = snprintf(NULL, 0, reply, strlen(userMessage), userMessage) + 1;
+    char *buffer = alloca(requiredSize);
+    snprintf(buffer, requiredSize, reply, strlen(userMessage), userMessage);
+
+    ssize_t write_result = write(socket, buffer, strlen(buffer));
+    (void)write_result;
+}
+
+int process_request(int new_socket, char *id) {
+    log_info("Processing HTTP SAML request\n");
+
+    int flag = 1;
+    setsockopt(new_socket, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int));
+
+    // Read the request
+    char request[1024];
+    ssize_t read_result = read(new_socket, request, sizeof(request));
+
+    // Check for '=id' in the response
+    // If the recevied request from the server is larger than the buffer, the result will not be null-terminated.
+    // Causing strlen to behave wrong.
+    if (read_result < 0 || read_result == sizeof(request)) {
+        log_error("Bad request\n");
+        send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+        return -1;
+    }
+
+    if (strlen(request) < 10 || strncmp(request, "GET /?id=", 9) != 0) {
+        log_error("Bad request\n");
+        send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+        return -1;
+    }
+
+    // Extract the id
+    const int id_start = 9;
+    char *id_end = memchr(&request[id_start], ' ', sizeof(request) - id_start);
+
+    if (id_end == NULL) {
+        log_error("Bad request format\n");
+        send_status_response(new_socket, "Invalid formatting of Fortinet server redirect response. VPN could not be established.");
+        return -1;
+    }
+
+    int id_length = id_end - &request[id_start];
+
+    if(id_length == 0 || id_length >= MAX_SAML_SESSION_ID_LENGTH) {
+        log_error("Bad request id\n");
+        send_status_response(new_socket, "Invalid SAML session id received from Fortinet server. VPN could not be established.");
+        return -1;
+    }
+
+    strncpy(id, &request[id_start], id_length);
```

Comment:
I'm confused here. `id_length` is determined by looking for a space. Then, you use this length with `strncpy`, which will _not_ insert a null terminator at `id[id_length-1]`. And yet, the data in `id` is expected to be null terminated, because it's used with `strlen` in `http.c`. If this code works, I guess it's because `id` happens to be initialized to zeros. But we can't rely on that.

I suggest using `memcpy` and adding a null at the end explicitly.

Reply:
This section was taken from the original author. I've replaced it with a completely different approach.